wallet: Pass through transaction locktime and preset input sequences and scripts to CreateTransaction

achow101 opened the pull request:
Currently `FundTransaction` handles transaction locktime and preset input data by extracting the selected inputs and change output from `CreateTransaction`'s results. This means that `CreateTransaction` is actually unaware of any user desired locktime or sequence numbers. This can have an effect on whether and how anti-fee-sniping works.

This PR makes `CreateTransaction` aware of the locktime and preset input data by providing them to `CCoinControl`. `CreateTransasction` will then set the sequences, scriptSigs, scriptWItnesses, and locktime as appropriate if they are specified. This allows `FundTransaction` to actually use `CreateTransaction`'s result directly instead of having to extract the parts of it that it wants.

Additionally `FundTransaction` will return a `CreateTransactionResult` as `CreateTransaction` does instead of having several output parameters. Lastly, instead of using `-1` as a magic number for the change output position, the change position is changed to be an optional with no value set indicating no desired change output position (when provided as an input parameter) or no change output present (in the result).

achow101 committed:
wallet: Introduce and use PreselectedInput class in CCoinControl

Instead of having different maps for selected inputs, external inputs,
and input weight in CCoinControl, have a class PreselectedInput which
tracks stores that information for each input.

achow101 committed:
wallet: Set preset input sequence through coin control

achow101 committed:
wallet: Explicitly preserve transaction locktime in CreateTransaction

We provide the preset nLockTime to CCoinControl so that
CreateTransactionInternal can be aware of it and set it in the produced
transaction.

achow101 committed:
wallet: Explicitly preserve transaction version in CreateTransaction

We provide the preset nVersion to CCoinControl so that
CreateTransactionInternal can be aware of it and set it in the produced
transaction.

achow101 committed:
wallet: use optional for change position as an optional in CreateTransaction

Instead of making -1 a magic number meaning no change or random change
position, use an optional to have that meaning.

achow101 committed:
wallet: return CreatedTransactionResult from FundTransaction

Instead of using the output parameters, return CreatedTransactionResult
from FundTransaction in the same way that CreateTransaction does.
Additionally, instead of modifying the original CMutableTransaction, the
result from CreateTransactionInternal is used.

achow101 committed:
wallet: Explicitly preserve scriptSig and scriptWitness in CreateTransaction

When creating a transaction with preset inputs, also preserve the
scriptSig and scriptWitness for those preset inputs if they are provided
(e.g. in fundrawtransaction).

luke-jr commented on src/wallet/rpc/spend.cpp:
```diff
@@ -488,13 +487,13 @@ static std::vector<RPCArg> FundTxDoc(bool solving_data = true)
 }
 
-void FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& fee_out, int& change_position, const UniValue& options, CCoinControl& coinControl, bool override_min_fee)
+std::optional<CreatedTransactionResult> FundTransaction(CWallet& wallet, const CMutableTransaction& tx, const UniValue& options, CCoinControl& coinControl, bool override_min_fee)
```
Since this can never return a nullopt (errors throw), maybe just return `CreatedTransactionResult`?

luke-jr commented:
>Lastly, instead of using -1 as a magic number for the change output position, the change position is changed to be an optional with no value set indicating no desired change output position (when provided as an input parameter) or no change output present (in the result).

I'm not sure this is an improvement. It just adds complexity, for no benefit in this particular instance?

luke-jr commented on src/wallet/coincontrol.h:
```diff
@@ -61,82 +128,153 @@ class CCoinControl
     }
 
-    void Select(const COutPoint& output)
+    void Select(const COutPoint& output, std::optional<uint32_t> sequence = std::nullopt, std::optional<CScript> script_sig = std::nullopt, std::optional<CScriptWitness> script_witness = std::nullopt)
```
I think it would be better to return the `PreselectedInput` and have the caller use `SetSequence` etc directly.

achow101 replied:
Done

luke-jr commented on src/wallet/spend.cpp:
```diff
@@ -832,11 +845,33 @@ static std::optional<CreatedTransactionResult> CreateTransactionInternal(
+        uint32_t sequence = default_sequence;
+        if (coin_control.HasSequence(coin.outpoint)) {
+            sequence = coin_control.GetSequence(coin.outpoint);
+            // If an input as a preset sequence, we can't do anti-fee-sniping
```
Why not? As long as one input has a non-final sequence, it should still work.

(`DiscourageFeeSniping` would need its sanity checks adjusted to tolerate it, though)

luke-jr replied:
Where is this behaviour in the current logic? I thought presently it would do the anti-sniping, and then possibly (at worst) have that partially "broken" by subsequent changes?

achow101 replied:
It's more of a side effect of `FundTransaction` rather than explicit behavior. Currently, having preset sequences inherently means a preset locktime which would disable anti-fee-sniping. This is just due to how `FundTransaction` works, and `FundTransaction` is the only way that preset sequences can even be provided.

Until we update `DiscourageFeeSniping` to handle preset sequences and locktime, I would prefer that we have this check explicitly just to avoid any problems in the future with how and when preset sequences and locktimes can be set.

darosior replied:
Making sure i got this one right. `use_anti_fee_sniping` may only be `false` if this is called from `FundTransaction`. And `FundTransaction` diregards the value of the `nLockTime` that may be set here anyways.

achow101 replied:
Yes, `FundTransaction` currently throws away `nLockTime` thereby disabling anti-fee-sniping, and this preserves that behavior.

MarcoFalke commented:
```
SUMMARY: MemorySanitizer: use-of-uninitialized-value src/wallet/spend.cpp:1025:5 in wallet::CreateTransaction(wallet::CWallet&, std::__1::vector<wallet::CRecipient, std::__1::allocator<wallet::CRecipient> > const&, std::__1::optional<unsigned int>, bilingual_str&, wallet::CCoinControl const&, FeeCalculation&, bool)

achow101 committed:
wallet: Replace SelectExternal with SetTxOut

Instead of having a separate CCoinControl::SelectExternal function, we
can use the normal CCoinControl::Select function and explicitly use
PreselectedInput::SetTxOut in the caller. The semantics of what an
external input is remains.

MarcoFalke commented:
There is also a lint issue with trailing whitespace or so?

furszy reviewed (commented):
Code reviewed, left few nits and topics.

How is enforced that `SignTransaction` doesn't overwrite the provided inputs' scriptsig and witness data?. It's called after setting the data and, as far as can see, it isn't skipping not empty inputs.
Would be good to add test coverage for it.

achow101 commented:
> How is enforced that `SignTransaction` doesn't overwrite the provided inputs' scriptsig and witness data?. It's called after setting the data and, as far as can see, it isn't skipping not empty inputs.

It will ignore inputs that have a valid scriptSig and witness. `SignTransaction` calls `DataFromTransaction` which will perform a `VerifyScript` of the existing data. If this is successful, it will mark that input as complete (sets `SignatureData.complete`) which `ProduceSignature` checks before it does anything else.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -1188,47 +1188,25 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
-        tx.vout.insert(tx.vout.begin() + nChangePosInOut, tx_new->vout[nChangePosInOut]);
-    }
-
-    // Copy output sizes from new transaction; they may have had the fee
-    // subtracted from them.
-    for (unsigned int idx = 0; idx < tx.vout.size(); idx++) {
-        tx.vout[idx].nValue = tx_new->vout[idx].nValue;
```
I think you misunderstand what the original code was doing. It isn't preserving the original values, it's replacing the original values with the values that `CreateTransaction` produces. Originally `FundTransaction` is calling `CreateTranasction` and copying specific parts of the result into the original transaction and returning that modified original. The change I've made here is to just return the result of `CreateTransaction` directly, which matches the previous behavior because the earlier commits are preserving all of the components of the transaction that `FundTranasction` was preserving.

furszy replied:
oops, yeah..  nvm then.

achow101 committed:
coincontrol: Replace HasInputWeight with returning optional from Get

achow101 commented on src/wallet/coincontrol.h:
```diff
@@ -66,77 +92,85 @@ class CCoinControl
-        txout = ext_it->second;
+        if (!it->second.HasTxOut()) {
+            return false;
+        }
```
Done

achow101 commented on src/wallet/coincontrol.h:
```diff
@@ -163,6 +173,20 @@ class CCoinControl
+    uint32_t GetSequence(const COutPoint& outpoint) const
+    {
+        return m_selected.at(outpoint).GetSequence();
+    }
```
Done, also for input weight and position.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -1136,15 +1136,15 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
+            } else {
+                error = _("Unable to find UTXO for external input");
+                return false;
+            }
```
Done as suggested

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -961,7 +961,15 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
+            script_witness = sw;
+        }
+        txNew.vin.push_back(CTxIn(coin.outpoint, script_sig, sequence));
+        txNew.vin.back().scriptWitness = script_witness;
```
Right now, I don't think that's a good idea. Since we are always say that the preset inputs have scriptSigs and witnesses when they come from `FundTransaction`, any input weight calculation using them may be incorrect because those inputs are not necessarily complete or valid, so the final weight may actually differ.

furszy reviewed (approved):
Code review ACK 91e6fad8

furszy reviewed (approved):
diff re-ACK 35fe41f

darosior reviewed (commented):
Concept and approach ACK. The code also looks good to me but i'm not too familiar with it so i may be missing some subtleties.

Since we now may have timelocked coins, another motivation for making `CreateTransaction` aware of a user chosen `nLockTime` would be to only select coins that are actually spendable with this `nLockTime` value.

theStack reviewed (commented):
Concept ACK

theStack commented on src/wallet/spend.cpp:
```diff
@@ -161,15 +161,17 @@ util::Result<PreSelectedInputs> FetchSelectedInputs(const CWallet& wallet, const
     const bool can_grind_r = wallet.CanGrindR();
     for (const COutPoint& outpoint : coin_control.ListSelected()) {
-        int input_bytes = -1;
+        int32_t input_bytes = coin_control.GetInputWeight(outpoint).value_or(-1);
```
(in commit 077783d24b7f1e3d3312e6cc12aee3f178400725)
Looking at the "override calculated size with coin control specified size" branch removed below, seems like a `GetVirtualTransactionSize` call is missing here (in the case that `GetInputWeight` doesn't return std::nullopt), as we want vbytes as `input_bytes` unit, not WU?

achow101 replied:
Good catch, fixed.

Tried to add a test but the effects of this mistake are not readily apparent enough to be tested.

theStack commented on src/wallet/coincontrol.cpp:
```diff
@@ -88,6 +88,20 @@ std::optional<uint32_t> CCoinControl::GetSequence(const COutPoint& outpoint) con
+bool CCoinControl::HasScripts(const COutPoint& outpoint) const
+{
+    const auto it = m_selected.find(outpoint);
+    if (it == m_selected.end()) {
+        return false;
+    }
+    return it->second.HasScripts();
```
(in commit 6fcbf9f44358f85aab4c0b6617c113b4c37da5a8)
nit: could be written as a one-liner
```suggestion
    return (it != m_selected.end() && it->second.HasScripts());
```

achow101 replied:
Done

achow101 commented on src/wallet/coincontrol.cpp:
```diff
@@ -14,69 +14,110 @@ CCoinControl::CCoinControl()
+    if (it == m_selected.end()) {
+        return false;
+    }
+    return it->second.HasTxOut();
```
Done

achow101 commented on src/wallet/coincontrol.cpp:
```diff
@@ -79,6 +79,15 @@ std::optional<int64_t> CCoinControl::GetInputWeight(const COutPoint& outpoint) c
+    if (it == m_selected.end()) {
+        return std::nullopt;
+    }
+    return it->second.GetSequence();
```
Done

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -1004,11 +1004,19 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
-        txNew.vin.push_back(CTxIn(coin->outpoint, CScript(), nSequence));
+        std::optional<uint32_t> sequence = coin_control.GetSequence(coin->outpoint);
+        if (sequence) {
+            // If an input as a preset sequence, we can't do anti-fee-sniping
```
Done

josibake commented:
big Concept ACK

Still in the process of reviewing this, but it looks good so far. Building off of this PR in https://github.com/bitcoin/bitcoin/pull/25273

theStack reviewed (approved):
Code-review ACK 59c363ec7245521e388b4d21abb44d7881907757

S3RK commented on src/wallet/coincontrol.h:
```diff
@@ -91,6 +91,8 @@ class CCoinControl
     //! SigningProvider that has pubkeys and scripts to do spend size estimation for external inputs
     FlatSigningProvider m_external_provider;
+    //! Locktime
+    std::optional<uint32_t> m_locktime;
```
in 03420020b6b4958ee6765be718724b89cd174d5c

`locktime` doesn't belong to `CoinControl` because it has nothing to do with coins being spent. I'd prefer if we keep `CoinControl` scoped down to only containing data necessary to select and spend coins.

Same is true for some other fields in `CoinControl`. Maybe we should extract parameters not related to coins into a separate struct?

achow101 replied:
With miniscript, it does matter since scripts can contain locktimes.

S3RK commented on src/wallet/coincontrol.h:
```diff
@@ -93,6 +93,8 @@ class CCoinControl
     //! Locktime
     std::optional<uint32_t> m_locktime;
+    //! Version
+    std::optional<uint32_t> m_version;
```
in c8544398fb945b2402274981b22c70fdd8a67b9e

Same comment as for `locktime`

achow101 replied:
With miniscript, it does matter, as it controls whether something with `OP_CSV` is spendable.

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -1347,15 +1347,15 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
+
             // The input was not in the wallet, but is in the UTXO set, so select as external
-            coinControl.SelectExternal(outPoint, coins[outPoint].out);
+            preset_txin.SetTxOut(coins[outPoint].out);
```
I'm not sure about 49ce0fe1e2ec87508aea537d17e6bb13791be1aa and probably would prefer to drop it.

It makes interface less clear. It's not obvious that I need to call `SetTxOut` to make input as external.

achow101 replied:
I'd prefer to leave it as is since it follows the pattern established by the following commits where various things are `Set`.

S3RK commented on src/wallet/coincontrol.h:
```diff
@@ -24,6 +24,33 @@ const int DEFAULT_MAX_DEPTH = 9999999;
 //! Default for -avoidpartialspends
 static constexpr bool DEFAULT_AVOIDPARTIALSPENDS = false;
 
+class PreselectedInput
```
in 59a7c7416313693a3c8219363d9ffee1e2b86d4d

Naming is hard. We already have struct `PreSelectedInputs` which is quite confusing as it refers to different data.

`PreSelectedInputs` is basically `set<COutput>`. It seems we should find a better name for `PreSelectedInputs`. Maybe rename it to `PreSelectedCoins`?

It'd be nice to have consistent terms to refer to different data.
How about we use `input` when we talk about tx.in and related info.
And we use `coin` when its enriched data in context of coin selection.

Perhaps @murchandamus have thought about terminology.

achow101 replied:
:shrug:

S3RK commented on src/wallet/coincontrol.h:
```diff
@@ -33,6 +33,10 @@ class PreselectedInput
     //! The sequence number for this input
     std::optional<uint32_t> m_sequence;
+    //! The scriptSig for this input
+    std::optional<CScript> m_script_sig;
```
in 53cb734591cabb37fc6e3d147a4715203cf15ddb

There is redundancy between `m_script_sig ` + `m_script_witness ` and `m_weight`. One can set all of them and reach and inconsistent state, which will lead to error.

Can we make the interface better?

achow101 replied:
It is actually expected that both can be provided, and not match. We may be funding a transaction with other inputs that already have some signatures (e.g. a multisig where one of the parties is adding another input for fees, and one other party has already signed with sighash_single), so `m_script_sig` or `m_script_witness` may already exist, but are incomplete. `m_weight` may then also be provided, and it wouldn't match since it would account for the signatures that have not yet been added.

S3RK reviewed (commented):
Approach ACK.

Left inline comments with more details.

It also would be nice to update PR description to explain what defects it fixes, i.e. expand "This can have an effect on whether and how anti-fee-sniping works." by describing potential bugs or negative effects.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -977,6 +977,10 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
     CMutableTransaction txNew; // The resulting transaction that we make
 
+    if (coin_control.m_version) {
+        txNew.nVersion = coin_control.m_version.value();
```
It isn't. When `FundTransaction` extracts the inputs and outputs from `CreateTransction`'s result and puts them into the tx it got, so the version ends up being preserved. But `CreateTransaction` isn't aware of the version selected by the user.

ishaanam commented on src/wallet/spend.cpp:
```diff
@@ -1346,50 +1391,31 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
-            coinControl.SelectExternal(outPoint, coins[outPoint].out);
+            preset_txin.SetTxOut(coins[outPoint].out);
         }
+        preset_txin.SetSequence(txin.nSequence);
```
This is done for every pre-selected input, even the ones that don't specify the sequence. This seems confusing and also means that a transaction with any pre-selected inputs won't do any anti-fee-sniping.

achow101 replied:
This is what currently happens. The behavior can be improved in future PRs.

ishaanam reviewed (commented):
Concept ACK

ishaanam commented on src/wallet/spend.cpp:
```diff
@@ -1341,6 +1346,9 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
         vecSend.push_back(recipient);
     }
 
+    // Set the user desired locktime
+    coinControl.m_locktime = tx.nLockTime;
```
In ebc39da8e5f72521fb4889bc1adc0a80136ce3da "wallet: Explicitly preserve transaction locktime in CreateTransaction": 

The user-provided locktime and sequence values should also be set in the `sendall` `coin_control`. In the future if we want to take locktime into account when running `AvailbleCoins`, this would be very useful.

achow101 replied:
That can be done in such future PRs.

ishaanam commented:
Looks like there is a silent merger conflict in `src/wallet/test/fuzz/notifications.cpp`

ishaanam reviewed (commented):
ACK 1d1a6ff03215cf3aadd5fd1118f354b1ba8ffa5a

theStack reviewed (approved):
re-ACK 1d1a6ff03215cf3aadd5fd1118f354b1ba8ffa5a

josibake commented on src/wallet/coincontrol.cpp:
```diff
@@ -114,4 +125,24 @@ std::optional<uint32_t> PreselectedInput::GetSequence() const
+    m_script_witness = script_wit;
+}
+
+bool PreselectedInput::HasScripts() const
```
in https://github.com/bitcoin/bitcoin/pull/25273/commits/2a06c4c2b33c614cc3f877bdc013c8df9e847ca9:

in earlier commits, you were replacing the "Has -> Get" pattern with a "Get" that returned a std::optional. Why can't we also do that here?

josibake commented on src/wallet/spend.cpp:
```diff
@@ -1162,7 +1162,13 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
-        txNew.vin.emplace_back(coin->outpoint, CScript(), sequence.value_or(default_sequence));
+        CScript script_sig;
+        CScriptWitness script_witness;
+        if (coin_control.HasScripts(coin->outpoint)) {
+            std::tie(script_sig, script_witness) = coin_control.GetScripts(coin->outpoint);
+        }
```
in https://github.com/bitcoin/bitcoin/pull/25273/commits/2a06c4c2b33c614cc3f877bdc013c8df9e847ca9:

could be something like:

```suggestion
        auto scripts = GetScripts(coin->outpoint);
        if (scripts) {
            std::tie(script_sig, script_witness) = *scripts;
        }
````

if you decide to drop the "Has" method.

josibake commented on src/wallet/coincontrol.cpp:
```diff
@@ -39,7 +39,9 @@ std::optional<CTxOut> CCoinControl::GetExternalOutput(const COutPoint& outpoint)
 {
-    return m_selected[output];
+    auto& input = m_selected[output];
+    input.SetPosition(m_selection_pos++);
```
in https://github.com/bitcoin/bitcoin/pull/25273/commits/e53789afe60ddbcc014931467ae6b4cab15d0c67:

TIL `m_var++` passes the value, then increments.

nit: I think it's considered more idiomatic (and more readable imo) to do something like:

```suggestion
    input.SetPosition(m_selection_pos);
    m_selection_pos++;
```

achow101 replied:
Done

josibake commented on src/wallet/interfaces.cpp:
```diff
@@ -281,12 +281,12 @@ class WalletImpl : public Wallet
     {
         LOCK(m_wallet->cs_wallet);
-        auto res = CreateTransaction(*m_wallet, recipients, change_pos,
+        auto res = CreateTransaction(*m_wallet, recipients, change_pos == -1 ? std::nullopt : std::make_optional(change_pos),
```
in https://github.com/bitcoin/bitcoin/pull/25273/commits/3d1a5f273c3c01b47ea4c74503c8353b2632dce8:

note for a follow-up: would nice if we could use this convention throughout the codebase instead of switching the magic number into a std::optional, like we are doing here.

josibake reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/25273/commits/1d1a6ff03215cf3aadd5fd1118f354b1ba8ffa5a

Left some nits that can be ignored or addressed in a follow-up, but nothing major. There is a type mismatch with `GetInputWeight` returning int64_t but being assigned to int32_t. Would be great to fix that but also not a super big deal since weight will never be greater than int32_t anyway (but should def be fixed in a followup).

achow101 commented on src/wallet/coincontrol.cpp:
```diff
@@ -14,69 +14,104 @@ CCoinControl::CCoinControl()
-    const auto ext_it = m_external_txouts.find(outpoint);
-    if (ext_it == m_external_txouts.end()) {
+    const auto it = m_selected.find(outpoint);
+    if (it == m_selected.end() || !it->second.HasTxOut()) {
```
The consequence is that size estimation will overestimate that input's size by 1 byte, which is up to 4 weight units of overestimation and therefore fees. However it should not be possible to accidentally set an internal input to be external as all calls to `SelectExternal` are guarded by a `!IsMine()`.

achow101 commented on src/wallet/coincontrol.cpp:
```diff
@@ -14,69 +14,104 @@ CCoinControl::CCoinControl()
+    return !m_selected.empty();
 }
 
 bool CCoinControl::IsSelected(const COutPoint& output) const
```
Renamed to outpoint here and elsewhere.

achow101 commented on src/wallet/coincontrol.cpp:
```diff
@@ -14,69 +14,104 @@ CCoinControl::CCoinControl()
 std::vector<COutPoint> CCoinControl::ListSelected() const
 {
-    return {m_selected_inputs.begin(), m_selected_inputs.end()};
+    std::vector<COutPoint> out;
```
Done

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -117,8 +117,9 @@ static std::optional<int64_t> GetSignedTxinWeight(const CWallet* wallet, const C
     // If weight was provided, use that.
-    if (coin_control && coin_control->HasInputWeight(txin.prevout)) {
-        return coin_control->GetInputWeight(txin.prevout);
+    std::optional<int32_t> weight;
```
`GetTransactionInputWeight` returns an `int64_t`, so I've unified these to use `int64_t` as well.

josibake commented:
ACK https://github.com/bitcoin/bitcoin/commit/0295b44c257fe23f1ad344aff11372d67864f0db

Thanks for taking the suggestions! Looks like the CI failure is a transient failure, I ran the same test locally and it passed. Probably just needs to be re-run.

S3RK commented:
Code review ACK 0295b44c257fe23f1ad344aff11372d67864f0db

Still not a huge fan of 596642c5a9 and I have same concerns as josibake. I prefer explicit defitino of external inputs rather than indirectly through `SetTxOut`. Though I don't want this to block the PR.

Also it seems two commits were squashed "Explicitly preserve scriptSig and scriptWitness in CreateTransaction" and "Preserve preset input order". Was that intentional?

josibake commented:
> Still not a huge fan of https://github.com/bitcoin/bitcoin/commit/596642c5a9f52dda2599b0bde424366bb22b3c6e and I have same concerns as josibake

My concerns were addressed by https://github.com/bitcoin/bitcoin/pull/25273#discussion_r1420969629. I think the approach in https://github.com/bitcoin/bitcoin/commit/596642c5a9f52dda2599b0bde424366bb22b3c6e is much simpler and as the comment points out, the worst-case scenario if inputs are labeled incorrectly is overestimating the fee by 1 byte.

josibake commented:
For the reviewers of this PR, https://github.com/bitcoin/bitcoin/pull/28560 is a follow-up that further refactors `FundTransaction` by cleaning up the SFFO parsing and removing the SFFO logic from `FundTransaction`. It does this by passing a `CRecipient` vector to `FundTransaction`, which also sets us up to be able to pass silent payment addresses to `FundTransaction` in a later PR.

furszy commented on src/wallet/spend.cpp:
```diff
@@ -1349,6 +1353,9 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
     coinControl.m_locktime = tx.nLockTime;
 
+    // Set the user desired version
+    coinControl.m_version = tx.nVersion;
```
nit:
it seems that the version field is not sanitized. The user provided `tx.nVersion` is a int32_t while `coinControl.m_version` is a uint32_t.
